server/api/src/rsDataObjLseek.cpp

```diff
@@ -10,112 +7,102 @@
     }
 
-    dataObjInfo = L1desc[l1descInx].dataObjInfo;
+    auto* dataObjInfo = l1desc.dataObjInfo;
 
-    // =-=-=-=-=-=-=-
-    // extract the host location from the resource hierarchy
+    // Extract the host location from the resource hierarchy.
     std::string location;
-    irods::error ret = irods::get_loc_for_hier_string( dataObjInfo->rescHier, location );
-    if ( !ret.ok() ) {
-        irods::log( PASSMSG( "rsDataObjLseek - failed in get_loc_for_hier_string", ret ) );
-        return -1;
+    if (const auto ret = irods::get_loc_for_hier_string(dataObjInfo->rescHier, location); !ret.ok()) {
+        irods::log(PASSMSG("rsDataObjLseek: failed in get_loc_for_hier_string", ret));
+        return ret.code();
     }
 
-
-    if ( getStructFileType( dataObjInfo->specColl ) >= 0 ) {
-        subStructFileLseekInp_t subStructFileLseekInp;
-        memset( &subStructFileLseekInp, 0, sizeof( subStructFileLseekInp ) );
+    if (getStructFileType(dataObjInfo->specColl) >= 0) {
+        subStructFileLseekInp_t subStructFileLseekInp{};
         subStructFileLseekInp.type = dataObjInfo->specColl->type;
-        subStructFileLseekInp.fd = L1desc[l1descInx].l3descInx;
+        subStructFileLseekInp.fd = l1desc.l3descInx;
         subStructFileLseekInp.offset = dataObjLseekInp->offset;
         subStructFileLseekInp.whence = dataObjLseekInp->whence;
-        rstrcpy( subStructFileLseekInp.addr.hostAddr,
-                 location.c_str(),
-                 NAME_LEN );
-        rstrcpy( subStructFileLseekInp.resc_hier,
-                 dataObjInfo->rescHier,
-                 NAME_LEN );
-        status = rsSubStructFileLseek( rsComm, &subStructFileLseekInp, dataObjLseekOut );
+        rstrcpy(subStructFileLseekInp.addr.hostAddr, location.c_str(), NAME_LEN);
+        rstrcpy(subStructFileLseekInp.resc_hier, dataObjInfo->rescHier, NAME_LEN);
+        return rsSubStructFileLseek(rsComm, &subStructFileLseekInp, dataObjLseekOut);
     }
-    else {
-        *dataObjLseekOut = ( fileLseekOut_t* )malloc( sizeof( fileLseekOut_t ) );
-        memset( *dataObjLseekOut, 0, sizeof( fileLseekOut_t ) );
-
-        ( *dataObjLseekOut )->offset = _l3Lseek( rsComm, l3descInx,
-                                       dataObjLseekInp->offset, dataObjLseekInp->whence );
-
-        if ( ( *dataObjLseekOut )->offset >= 0 ) {
-            status = 0;
-        }
-        else {
-            status = ( *dataObjLseekOut )->offset;
-        }
+
+    // If the replica was opened in read-only mode, then don't allow the client
+    // to seek past the data size in the catalog. This is necessary because the
+    // size of the replica in storage could be larger than the data size in the
+    // catalog.
+    //
+    // For all other modes, let the seek operation do what it normally does.
+    const auto offset = (O_RDONLY == (l1desc.dataObjInp->openFlags & O_ACCMODE))
+        ? std::min(dataObjLseekInp->offset, dataObjInfo->dataSize)
+        : dataObjLseekInp->offset;
```

Comment:
Moved the offset adjustment below the special collections code. CI tests involving special collections were failing.

It is probably best to ignore special collections for this anyway. Thoughts?

Reply:
you're only adding a new safety check - certainly not changing the current behavior for them until we swing around again and tackle them all head on seems fine.  @alanking ?

Reply:
I have found ignoring special collections and/or pushing them to the fringe cases is a good policy, so seems fine to me.

Reply:
Ok. Marking as resolved.